utility/getPipelineStages.py

```diff
@@ -0,0 +1,128 @@
+import json
+import logging
+import os
+import random
+import string
+import sys
+
+import yaml
+from docopt import docopt
+
+log = logging.getLogger(__name__)
+doc = """
+This script fetches all the tests to be run for a pipeline based on the RHCS version and overrides
+
+    Usage:
+        utility/getPipelineStages.py --rhcephVersion <VER> --tags <tags>
+                  [--overrides <str>]
+                  [--metadata <FILE>]
+
+        utility/getPipelineStages.py (-h | --help)
+
+    Options:
+        -h --help          Shows the command usage
+        -v --rhcephVersion VER     The rhcephVersion for which test stages need to be fetched
+        -t --tags <str>    tags to be used for filtering test scripts
+        -o --overrides <str>       Overrides to be considered for execution
+        -m --metadata FILE         Metadata file full path if it is not present in default pipeline/metadata folder
+"""
+
+
+def generate_random_string(length):
+    """Generate a random alphanumeric string of given length"""
+    return "".join(random.choices(string.ascii_uppercase + string.digits, k=length))
+
+
+def fetch_stages(args):
+    """
+    Fetch test stages to be executed based on the input provided
+    Args:
+        args:
+            Dictionary containing the tier_level, cloud_type and all other necessary override parameters
+            to be used to fetch test stages for execution
+
+            arguments = {
+                "rhcephVersion": <the rhbuild version for which the stages need to be fetched>,
+                "tags": comma separated string containing the tags based on which the stages should be filtered,
+                "overrides": <overrides in json format, all keys accepted by run.py are supported to be overridden
+                            if any key similar to --store which does not have a value it can be passed as "store": ""
+                "metadata_file": <path to the metadata file>,
+            }
+
+            Example:
+                args = {
+                    "rhcephVersion": 5.1,
+                    "tags": "tier-1,stage-1,ibmc",
+                    "overrides": {"cloud_type": "psi", "platform": "rhel-8"},
+                    "metadata_file": "pipeline/metadata/5.1_custom.yaml"
+
+    Returns:
+        Yaml string containing CLI to be executed for each test suite at every stage
+        Examples:
+            "test-cephfs-core-features":
+                "execute_cli": ".venv/bin/python run.py ......",
+                 "cleanup_cli": "...."
+            .....
+    """
+    rhceph_version = args.get("rhcephVersion")
```

Comment:
As both the parameters are not mandatory and `metadata_file` is depends on `rhceph_version` parameter, below code could be better -
```
    metadata_file = args.get("metadata_file")
    if not metadata_file and args.get("rhcephVersion"):
    	metadata_file = f"pipeline/metadata/{rhceph_version}.yaml
    else
        rase Exception(
        	"Mandatory parameter 'rhcephVersion' and "
        	"'metadata' file are not specified")
```

Reply:
Few mistakes in the code -
1. If condition could be - `if not metadata_file and "rhcephVersion" in args.keys()`
2. `rhceph_version` is parameter which is missing and can be fixed as -  `f"pipeline/metadata/{}.yaml".format(args.get("rhcephVersion"))`

Reply:
But the above condition will throw an exception for the scenario where metadata file is present and rhcephVersion is not present, which is not desired. It should actually just use the given metadata file in that scenario. 

So I guess the below code snippet would do the job.

metadata_file = args.get("metadata_file")
rhceph_version = args.get("rhcephVersion")
if not metadata_file and rhceph_version:
    metadata_file = f"pipeline/metadata/{rhceph_version}.yaml
elif not metadata_file and not rhceph_version:
    raise Exception(
     "Cannot fetch test stages since rhcephVersion and metadata file are not specified")

Reply:
Also as per the recent discussion I had with @psathyan we are removing the concept of passing metadata file and expecting that the metadata will definitely be in pipeline/metadata/<rhcephVersion>.yaml path.